api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/authenticate/ForgotPasswordController.java

```diff
@@ -0,0 +1,416 @@
+            return url.toExternalForm();
+        } catch (MalformedURLException e) {
+            return "error_creating_password_link";
+        }
+    }
+
+    /**
+     * Calculates the expiration date for a password reset link.
+     *
+     * @return The expiration date as a Date object.
+     */
+    private Date calculateExpirationDate() {
+        Calendar c = Calendar.getInstance();
+        c.add(Calendar.DATE, DAYS_TO_USE_PASSWORD_LINK);
+        return c.getTime();
+    }
+
+    /**
+     * Retrieves the URL for the "Forgot Password" page from the VitroRequest context.
+     *
+     * @param request The VitroRequest object containing request information.
+     * @return The URL for the "Forgot Password" page as a string.
+     */
+    private String getForgotPasswordUrl(VitroRequest request) {
+        String contextPath = request.getContextPath();
+        return contextPath + "/forgot-password";
+    }
+
+    /**
+     * Retrieves the URL for the "Contact" page from the VitroRequest context.
+     *
+     * @param request The VitroRequest object containing request information.
+     * @return The URL for the "Contact" page as a string.
+     */
+    private String getContactUrl(VitroRequest request) {
+        String contextPath = request.getContextPath();
+        return contextPath + "/contact";
+    }
+
+    /**
+     * Checks whether the functionality for password recovery is enabled based on the configuration.
+     *
+     * @return 'true' if the functionality is enabled, 'false' otherwise.
+     */
+    private boolean isFunctionalityEnabled() {
+        String enabled = ConfigurationProperties.getInstance().getProperty("authentication.forgotPassword");
+        return enabled != null && enabled.equalsIgnoreCase("enabled");
+    }
+
+    /**
+     * Causes the current thread to sleep for a random duration to mitigate time-based attacks.
+     */
+    private void sleepForRandomTime() {
+        Random random = new Random();
+        int minSleepTime = 100;
+        int maxSleepTime = 500;
+        int randomSleepTime = random.nextInt(maxSleepTime - minSleepTime + 1) + minSleepTime;
+
+        try {
+            Thread.sleep(randomSleepTime);
+        } catch (InterruptedException e) {
+            log.error(randomSleepTime + "ms sleep time for mitigating time-based attacks was interrupted.");
+        }
+    }
+
+    /**
+     * Retrieves the value of the specified request parameter and ensures it is non-null.
+     * If the parameter value is null, an empty string is returned.
+     * Leading and trailing whitespace is removed from the parameter value.
+     *
+     * @param req the HttpServletRequest containing the parameters
+     * @param key the key of the parameter to retrieve
+     * @return the non-null and trimmed value of the specified request parameter,
+     * or an empty string if the parameter is null
+     */
+    private String nonNullAndTrim(HttpServletRequest req, String key) {
```

Comment:
People sometimes accidentally add space characters at the end of an input field, using trim will add some quality of life in my opinion as it will not treat that as some kind of invalid input.

Reply:
Maybe name it something like getTrimmedParameterValue ?